File: dictionary/verify_alien_dictionary.py
"""
https://leetcode.com/problems/verifying-an-alien-dictionary
Time: O(N)
Space: O(1)
"""

from typing import List


class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        if not words:
            return False
        
        lookup = { char : index for index, char in enumerate(order)}
        
        for i in range(len(words)-1):
            for j in range(len(words[i])):
                
                if j >= len(words[i+1]):
                    return False
                
                if words[i][j] != words[i+1][j]:
                    if lookup[words[i][j]] > lookup[words[i+1][j]]:
                        return False
                    break
        
        return True

File: dictionary/verify_alien_dictionary.java
"""
https://leetcode.com/problems/verifying-an-alien-dictionary
Time: O(N * M)
Space: O(1)
"""
class Solution {
    int[] lookup = new int[26];
    public boolean isAlienSorted(String[] words, String order) {
        
        for(int index = 0; index < order.length(); index++){
            lookup[order.charAt(index)-'a'] = index;
        }
        
        for(int index = 1; index < words.length; ++index){            
            if(compare(words[index-1], words[index])){
                return false;  
            } 
        }
        
        return true;
    }
    
    private boolean compare(String prev, String current){
        int m = prev.length();
        int n = current.length();
        
        for(int index = 0; index < m && index < n; index++){
            if(prev.charAt(index) != current.charAt(index)){
                return lookup[prev.charAt(index)-'a'] > lookup[current.charAt(index)-'a'];    
            }
        }
        
        return m > n;
    }
}